### PR TITLE
CB-9563 Mulptipart form data is used even a header named Content-Type…

### PR DESCRIPTION
**Pull request by daserge**
… is present

Adds direct upload endpoint

[Jira issue](https://issues.apache.org/jira/browse/CB-9563)

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -35,15 +37,34 @@ http.createServer(function (req, res) {
...
-            console.log(stringify({fields: fields, files: files}));
+        if(req.headers["content-type"].indexOf("multipart/form-data") === 0) {
+            console.log("multipart/form upload");
+            var form = new formidable.IncomingForm();
+            form.parse(req, function(err, fields, files) {
+                res.writeHead(200, {'content-type': 'text/plain'});
+                console.log(stringify({fields: fields, files: files}));
+
+                res.write(stringify({fields: fields, files: files}));
+                console.log
```
Comment (dblotsky) [inline server.js]: Empty `console.log` here, is it intentional?

File: server.js [hunk 2]
```diff
@@ -4,6 +4,8 @@ var formidable = require('formidable'),
     port = process.env.PORT || 5000;
     stringify = require('json-stringify-safe');
 
+var DIRECT_UPLOAD_LIMIT = 32;
```
Comment (dblotsky) [inline server.js]: What are the units for this?

Comment (daserge) [inline server.js]: It's bytes; this corresponds to the test: https://github.com/apache/cordova-plugin-file-transfer/pull/117/files#diff-2a8a5fef3397df87ab538f028a5c6b50R955 and the contents: https://github.com/apache/cordova-plugin-file-transfer/blob/9e93bad83c9e980cb593dc057fe8af40f35652f1/tests/tests.js#L710, which will be 17 bytes.

Comment (dblotsky) [inline server.js]: Please add that in a comment.

File: server.js (hunk 1 above)
Comment (daserge) [inline server.js]: Thanks, I will remove it. It was there before though, but it makes no sense.

File: server.js (hunk 2 above)
Comment (dblotsky) [inline server.js]: I.E. The fact that it's bytes.

**Comment by daserge** [comment]
@dblotsky, updated.

**Comment by dblotsky** [comment]
Do you think we can have some tests for this file? It's tedious to test it manually with `curl`. :/

**Comment by dblotsky** [comment]
Also... hold on... this server actually writes the files to disk. It should *not* do that.

**Comment by daserge** [comment]
I usually test it locally or with Heroku along with cordova-plugin-test-framework and the plugin auto tests.
What do you mean by writing to disk? Can you please point to the corresponding code?

**Comment by dblotsky** [comment]
The `formidable` package writes `multipart-form` files to a temporary file on disk. This is a **huge** security problem.

**Comment by dblotsky** [comment]
Please rewrite the form code so that nothing gets written to disk.

**Comment by daserge** [comment]
@dblotsky, @stevengill, @shazron - Can you please advice on how to handle this issue?
Should we switch to manual form parsing or to some lib supporting in-memory storing?
I'm also not sure how the security issues need be tracked - should I create a Jira for this?

**Comment by dblotsky** [comment]
We were only running this on one machine, and we took that machine down and requested for it to be rebuilt, so no worries. As for the change, we should either look into configuring Formidable not to write to disk, or into using another library. Manual parsing would be the least favourable option.

**Comment by daserge** [comment]
Created https://issues.apache.org/jira/browse/CB-10080 for the disk-cache issue.

**Pull request merged**